### PR TITLE
Improve CI path-changes-based-triggers to work better

### PR DESCRIPTION
**Pull request by radical**
## Background:

`eng/pipelines/common/evaluate-default-paths.yml` is used to specify named subsets of paths, and `evaluate-changed-paths.sh` decides which subsets have "changed files". And these subsets are used in conditions for the various jobs to determine when they should be run.

---
`evaluate-changed-paths.sh: Add support for include+exclude combined`

## Problem

This script currently supports `--include`, and `--exclude`
parameters to determine if we have any changed paths.

```
Scenarios:
1. exclude paths are specified
    Will include all paths except the ones in the exclude list.
2. include paths are specified
    Will only include paths specified in the list.
1st we evaluate changes for all paths except ones in excluded list. If we can not find
any applicable changes like that, then we evaluate changes for included paths
if any of these two finds changes, then a variable will be set to true.
```

As described above, these two are evaluated exclusively.
For example:

```
include: [ '/a/b/*' ], exclude: [ '/a/b/d.txt' ]
```

- This would return true if there are changes:
  - `[ '/a/b/c.txt' ]` - caught by `include`
  - or `[ '/r/s.txt' ]` - caught by `exclude`

- but it would also return true for `[ '/a/b/d.txt' ]` because `include`
does not consider the `exclude` list.

## Solution:

- This commit adds a new `--combined` parameter which essentially supports
`$include - $exclude`. Thus allowing exact more constrained path
specification.

- It is passed in addition to `include`, and `exclude`.

Given:
```
include: [ '/a/b/*' ], exclude: [ '/a/b/d.txt' ]
```

- For the following path changes:
  - `[ '/a/b/c.txt' ]` - `true`
  - `[ '/r/s.txt' ]` - `false` because this does not fall in
    `$include - $exclude`
  - `[ '/a/b/d.txt' ]` - `false` - excluded in `$include - $exclude`

The implementation is trivially implemented by passing both the lists to
`git diff` which supports this already.

---

- Track subset name changes in the ymls
- Update `wasm` jobs to have tighter trigger conditions

- This results in wasm specific changes only triggering relevant wasm jobs. For example, if there are wasm debugger changes then only the wasm debugger jobs will be triggered. Or if there are only workload manifest[1] changes then Wasm.Build.Tests will be triggered only.

### CODE REVIEW

**Comment by radical** [comment]
reviving an old branch

**Comment by radical** [comment]
cc @mmitche @MichalStrehovsky @fanyang-mono @directhex

**Comment by radical** [comment]
Reviewing individual commits should be easier.

**Comment by radical** [comment]
Any objections to this?

**Comment by radical** [comment]
> I haven't reviewed the actual code yet, but I wonder why there needs to be a new `--combined` parameter. The way you described it, it sounds like the `--combined` behavior is the expected behavior, and the behavior we have now is unintuitive.

True. I was expecting the existing behavior to be like `--combined`, but that wasn't the case. And the various job conditions seem to depend on this difference too. Since, we don't really have any kind of automated tests for this I'm taking the slow approach, and changing only what I can test myself - (mostly) wasm builds, and can track.

In the past with similar changes, we had situations where triggers for some non-wasm builds got disabled as an unintended side-effect, and I'm not familiar with all the builds to be 100% sure here.

> Do we want to just take a breaking change and remap as necessary?

Maybe we can take the extra parameter for now. And the various area owners can gradually move their builds to this, and then eventually we make this the default:) The `only include`, and `only exclude` cases could be expressed with a default `--combined` too, by splitting the lists, probably reducing readability? I would prefer that the owners for those builds make the changes though!

**Comment by radical** [comment]
A tool could possibly be written that would read the subsets, and check the conditions for all the jobs against that. Check for things like the expected list of jobs being triggered by path changes, and dependent jobs also being triggered etc.

**Comment by ViktorHofer** [comment]
This shouldn't impact official builds but just to be sure, can we wait with merging this in until next week when we branched-off for P1? Dependency flow is already complicated enough and I believe this change falls under the "general infrastructure improvement" bucket that isn't necessary to merge in for Preview 1. Does that make sense?

**Comment by radical** [comment]
Yes, this can wait for the branch. Thank you for taking a look 👍

**Comment by radical** [comment]
ping @akoeplinger @ViktorHofer @trylek @agocke for review

File: eng/pipelines/runtime-official.yml [hunk 1]
```diff
@@ -481,6 +481,10 @@ extends:
       #
       # Build Workloads
       #
+      # note: if this is changed to use evaluatePaths job,
```
Comment (akoeplinger) [inline eng/pipelines/runtime-official.yml:484]: the official build doesn't do path evaluation by design so I don't think we need this comment here

**Review by akoeplinger — APPROVED** [review]
LGTM with one comment

**Comment by radical** [comment]
The `Evaluate paths` job is running successfully on the pipelines. And there are no yml errors on azdo.
Since this is a CI only change, running all the jobs won't add any value, I'll merge this now.

**Pull request merged**